src/ConversationList/ConversationListItem.vala

```diff
@@ -32,12 +32,36 @@ public class Mail.ConversationListItem : Gtk.ListBoxRow {
 
     public int64 timestamp {
         get {
+            int64 timestamp = 0;
             if (node.message.date_received == 0) {
                 // Sent messages do not have a date_received timestamp.
                 return node.message.date_sent;
+                timestamp = node.message.date_sent;
+
+                unowned Camel.FolderThreadNode? child = (Camel.FolderThreadNode?) node.child;
+                while (child != null) {
+                    if (child.message.date_sent > timestamp) {
+                        timestamp = child.message.date_sent;
+                    }
+
+                    child = (Camel.FolderThreadNode?) child.next;
+                }
+
+                return timestamp;
             }
 
             return node.message.date_received;
+            timestamp = node.message.date_received;
+            unowned Camel.FolderThreadNode? child = (Camel.FolderThreadNode?) node.child;
+            while (child != null) {
```

Comment:
Instead of branching on whether `node.message.data_received != 0`, and then looping over all messages in the message thread, I think we should have a single loop through the message thread, and then calculate the `max(date_received, date_sent)` (pseudocode) of all messages.
Otherwise, if the thread starts with a sent message (= no `date_received`), the `date_received` timestamps of other messages in the thread will be ignored.